source/s3_client.c

```diff
@@ -352,6 +356,7 @@ struct aws_s3_client *aws_s3_client_new(
     struct aws_s3_buffer_pool_usage_stats pool_usage = aws_s3_buffer_pool_get_usage(client->buffer_pool);
 
     if (client_config->max_part_size > pool_usage.mem_limit) {
+        aws_s3_buffer_pool_destroy(client->buffer_pool);
```

Comment:
trying to clean up behind yourself is hard in C
which is why we prefer the "goto error, cleanup everything" pattern,
and why our cleanup/destroy functions are the only ones that tolerate unexpected NULL args

This function already has 2 different "goto error" labels.
Now 1 of those labels you have to be careful to also cleanup a specific variable before you goto it. this is too fragile 😓

Seems better to completely remove the `on_early_fail` label from this function, and just always try to cleanup everything via `goto on_error`?

Another option: instead of the `on_error:` label sharing 90% the same cleanup code as `s_s3_client_finish_destroy_default()`, have them share code, and tolerate destroying a half-assembled client

Reply:
Thanks, I have moved code around to get rid of `on_early_fail`.